pkg/filters/grpcproxy/pool.go

```diff
@@ -0,0 +1,485 @@
+	}
+
+	panic(fmt.Errorf("should not reach here"))
+}
+
+func (sp *ServerPool) doHandle(ctx stdcontext.Context, spCtx *serverPoolContext) error {
+	lb := sp.LoadBalancer()
+	svr := lb.ChooseServer(spCtx.req)
+	// if there's no available server.
+	if svr == nil {
+		logger.Debugf("%s: no available server", sp.name)
+		return serverPoolError{status.New(codes.InvalidArgument, "no available server"), resultClientError}
+	}
+	if f, ok := lb.(ReusableServerLB); ok {
+		defer f.ReturnServer(svr)
+	}
+	// maybe be rewrite by grpcserver.MuxPath#rewrite
+	fullMethodName := spCtx.req.FullMethod()
+	if fullMethodName == "" {
+		return serverPoolError{status.New(codes.InvalidArgument, "unknown called method from context"), resultClientError}
+	}
+	send2ProviderCtx, cancelContext := stdcontext.WithCancel(metadata.NewOutgoingContext(ctx, spCtx.req.RawHeader().GetMD()))
+	defer cancelContext()
+	dialCtx, cancel := stdcontext.WithCancel(stdcontext.Background())
+	if sp.spec.BlockUntilConnected {
+		dialCtx, cancel = stdcontext.WithTimeout(dialCtx, sp.connectTimeout)
+	}
+	defer cancel()
+	conn, err := grpc.DialContext(dialCtx, svr.URL, sp.dialOpts...)
+	if err != nil {
+		logger.Infof("create new conn without pool fail %s for source addr %s, target addr %s, path %s",
+			err.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+		return serverPoolError{status: status.Convert(err), result: resultInternalError}
+	}
+	proxyAsClientStream, err := conn.NewStream(send2ProviderCtx, desc, fullMethodName)
+	defer conn.Close()
+	if err != nil {
+		logger.Infof("create new stream fail %s for source addr %s, target addr %s, path %s",
+			err.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+		return serverPoolError{status: status.Convert(err), result: resultInternalError}
+	}
+
+	result := sp.biTransport(spCtx, proxyAsClientStream)
+	if result != nil && result != io.EOF {
+		logger.Infof("create new stream fail %s for source addr %s, target addr %s, path %s",
+			result.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+	}
+	return result
+}
+
+func (sp *ServerPool) biTransport(ctx *serverPoolContext, proxyAsClientStream grpc.ClientStream) error {
+	// Explicitly *do not Close* c2sErrChan and c2sErrChan, otherwise the select below will not terminate.
+	// Channels do not have to be closed, it is just a control flow mechanism, see
+	// https://groups.google.com/forum/#!msg/golang-nuts/pZwdYRGxCIk/qpbHxRRPJdUJ
+	c2sErrChan := sp.forwardE2E(ctx.stdr, proxyAsClientStream, nil)
+	s2cErrChan := sp.forwardE2E(proxyAsClientStream, ctx.stdw, ctx.resp.RawHeader())
+	// We don't know which side is going to stop sending first, so we need a select between the two.
+	state := stateNormal
+	for state != stateException {
+		select {
+		case c2sErr := <-c2sErrChan:
+			if c2sErr == io.EOF {
+				// this is the happy case where the sender has encountered io.EOF, and won't be sending anymore./
+				// the clientStream>serverStream may continue pumping though.
```

Comment:
```suggestion
				// this is the happy case where the sender has encountered io.EOF, and won't be sending anymore.
				// the clientStream --> serverStream may continue pumping though.
```

Reply:
for "the clientStream --> serverStream may continue pumping though", is it true? or should it be "the serverStream --> clientStream may continue pumping though"?

Reply:
it is confusing, so I replaced it with a clearer statement.
by the way, easegress use grpc.ClientStream to recv msg from server and send message to server, easegress use grpc.ServerStream to recv msg from client and send message to client.